### PR TITLE
netconf-1.1 support

### PR DESCRIPTION
**Pull request by martin-volf**
Support for 1.1 NETCONF framing. The change had to rework also the old 1.0 framing 1. because it was buggy 2. to make framing transparent to the user.

### CODE REVIEW

File: src/main/java/com/tailf/jnc/framing/Framer.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+
+
+/**
+ * Frame formatting and parsing.
+ */
+
+public interface Framer {
```
Comment (jomiklos) [inline src/main/java/com/tailf/jnc/framing/Framer.java:21]: shouldn't this be protected?

File: src/main/java/com/tailf/jnc/Element.java [hunk 2]
```diff
@@ -1717,6 +1717,44 @@ private void toXMLString(int indent, StringBuffer s) {
         s.append(getIndentationSpacing(flag, indent)).append("</").append(qName).append(">\n");
     }
 
+    public String encodedXMLString(boolean newline_at_end) {
+        final StringBuffer s = new StringBuffer();
+        encodedXMLString(s, newline_at_end);
+        return s.toString();
+    }
+
+    private void encodedXMLString(StringBuffer s, boolean newline_at_end) {
```
[the remark this answers is not in the document]
Comment (martin-volf) [reply src/main/java/com/tailf/jnc/Element.java:1726]: Fixed.

File: src/main/java/com/tailf/jnc/framing/Framer.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+@SuppressWarnings("PMD.ClassNamingConventions")
+class NC1_1_Framer extends BaseFramer {
+    static final int MAX_HEADER_SIZE = 13;
+    byte[] chunkHdr = new byte[MAX_HEADER_SIZE];
+
+    static final int LF;
+    static final int HASH;
+
+    static {
```
[the remark this answers is not in the document]
Comment (martin-volf) [reply src/main/java/com/tailf/jnc/framing/Framer.java:90]: Oh... now that was some over-engineering. Fixed.

File: src/main/java/com/tailf/jnc/framing/Framing.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    public Framer newSessionFramer(DataReader rdr, InputStream in, OutputStream out) {
+        try {
+            Constructor<? extends BaseFramer> cstr = clazz.getConstructor(DataReader.class,
+                                                                          InputStream.class,
+                                                                          OutputStream.class);
+            return cstr.newInstance(rdr, in, out);
+        } catch (ReflectiveOperationException e) {
+            // what?
```
[the remark this answers is not in the document]
Comment (martin-volf) [reply src/main/java/com/tailf/jnc/framing/Framing.java:31]: Fixed (somewhat; still no proper logging available).

File: src/main/java/com/tailf/jnc/framing/Framer.java (hunk 1 above)
Comment (martin-volf) [reply src/main/java/com/tailf/jnc/framing/Framer.java:21]: No? It is used elsewhere.

File: src/main/java/com/tailf/jnc/framing/README.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+* The 1.0 framing mechanism, on the other hand works best with String
+  instances, in particular because it is necessary to search the endmarker
+  string when receiving; with bytes, this is tricky to do correctly and not
+  natively supported by Java.  This also implies using `Reader`/`Writer`
+  instances.
+  
+The components are these, described top-down:
+
+* Enumeration `Framing` with two instances (for 1.0 and 1.1 framing mechanisms)
```
[the remark this answers is not in the document]
Comment (martin-volf) [reply src/main/java/com/tailf/jnc/framing/README.md:23]: Well, enum variants are instances of that enum, but it might be misleading indeed. Fixed.

**Review by jojohans — APPROVED** [review]

**Pull request merged**